mllib/src/main/scala/org/apache/spark/ml/tuning/CrossValidator.scala

```diff
@@ -117,6 +123,12 @@ class CrossValidator @Since("1.2.0") (@Since("1.4.0") override val uid: String)
     instr.logParams(numFolds, seed, parallelism)
     logTuningParams(instr)
 
+    val collectSubModelsParam = $(collectSubModels)
+
+    var subModels: Option[Array[Array[Model[_]]]] = if (collectSubModelsParam) {
```

Comment:
so this var seems unnecessary, could we just it seems like we'd be better by just collecting modelFutures in copy values (then we can avoid the mutation on L145)

Reply:
@holdenk @jkbradley I already thought about this issue. The reason I use this way is:
1) When `$(collectSubModels) == false`, the `modelFutures` and `foldMetricFutures` will be executed in pipelined way, this will make sure that the `model` generated in `modelFutures` will be released in time, so that the maximum memory cost will be `numParallelism * sizeof(model)`.  If we use the way of "collecting modelFutures", it will increase the memory cost to be `$(estimatorParamMaps).length * sizeof(model)` . This is a serious issue which is discussed before.
2) IMO the mutation on L145 won't influence performance. and it do not need something like lock, there is no race condition.

Reply:
I don't follow with #1, if we keep all the models (e.g. set `collectSubModelsParam`)  then the maximum memory cost will be `$(estimatorParamMaps).length * sizeof(model)` in either case? If we don't keep the models (e.g. set `collectSubModelsParam` to false) then you don't have to collect the future back at the end and there is no additional overhead.

For #2, It's not that mutation impacts performance, its that it makes the code less easy to reason about for no gain (unless I've misunderstood something about part 1).

Reply:
@holdenk Oh, sorry for confusing you. Yes, if set `collectSubModelsParam` the memory cost will always be `$(estimatorParamMaps).length * sizeof(model)`. According to your suggestion, we have to duplicate code logic (but if i am wrong correct me):
- When set `collectSubModelsParam`, we cannot pipeline `modelFutures` and `foldMetricFutures`, we should execute `modelFutures` and collect results first, and modify `foldMetricFutures` logic (change it into the way passing `model` param, not by `modelFuture.map { model => ...} ).
- When not set  `collectSubModelsParam`, just keep current `modelFutures` & `foldMetricFutures` and pipeline them to execute.
So, according to your suggestion, it seems need more code. So do you still prefer this way ? Or do you have better way to implement that ?

Reply:
Sorry I didn't follow up on this before.  I think that @WeichenXu123 's argument is valid, but please say if there are issues I'm missing @holdenk